File: dotnet/source/cmf.eventing/IEventHandler.cs
using System;
using System.Collections.Generic;
using cmf.bus;


namespace cmf.eventing
{
    /// <summary>
    /// Defines an interface to be implemented by types that wish to process received events.
    /// </summary>
    /// <seealso cref="IEventConsumer.Subscribe"/>
    public interface IEventHandler
    {
        /// <summary>
        /// The topic of the event which the handler is intended to receive.
        /// </summary>
        ///TODO: Why do we do topic here and type in Java???
        string Topic { get; }

        /// <summary>
        /// This method is invoked when an event of handled type is received. 
        /// It is the method that should handle the received event.
        /// </summary>
        /// <param name="ev"></param>
        /// <param name="headers"></param>
        /// <returns>An object indicating the outcome of handling the event.</returns>
        /// <remarks>How the return value is interpreted is dependent upon the 
        /// <see cref="IEventConsumer" /> implementation</remarks>
        object Handle(object ev, IDictionary<string, string> headers);

        /// <summary>
        /// This method is invoked when an exception occurs attempting to handle 
        /// an envelope that meets the criteria to be handled by this handler. 
        /// </summary>
        /// <param name="env"></param>
        /// <param name="ex"></param>
        /// <returns>An object indicating the outcome of handling the event.</returns>
        /// <remarks><para>How the return value is interpreted is dependent upon the 
        /// <see cref="IEventConsumer" /> implementation.</para>
        /// <para>The relationship between this method and the 
        /// <see cref="IRegistration.HandleFailed" />
        /// method is implementation dependent.</para></remarks>
        object HandleFailed(Envelope env, Exception ex);
    }
}

File: java/cmf-eventing-api/src/main/java/cmf/eventing/IEventHandler.java
package cmf.eventing;

import java.util.Map;

import cmf.bus.Envelope;

/**
 * Defines an interface to be implemented by types that which to process received events. 
 */
public interface IEventHandler<TEVENT> {

	/**
	 * Returns the type of the event which the handler is intended to receive.  
	 * Must be a non-abstract,  non-generic type.  Only events of the exact type
	 * will be received. I.e. events that are sub-types of the returned type 
 	 * will not be received.
	 * 
	 * @return The Class of the event to be handled.
	 */
    Class<TEVENT> getEventType();

    /**
     * The method that processes the received event.
     * 
     * @param event The event to be processed.
     * @param headers The {@link cmf.bus.Envelope} headers of the envelope in which
     * the event arrived. 
     * 
     * @return An object indicating the outcome of handling the event.
     */
    Object handle(TEVENT event, Map<String, String> headers);

    /**
     * This method will be invoked when an exception occurs attempting to handle an 
     * envelope that meets the registration and filter criteria.
     * 
     * @param env The envelope in which the event was received.
     * @param ex The exception that occurred.
     * 
     * @return An object indicating the outcome of handling the envelope.
     */
    Object handleFailed(Envelope envelope, Exception e);
    //TODO: This method is never called by the underlying implementation.
}
